[pdns] Add new role debops.pdns

imrejonk opened the pull request:
This adds debops.pdns, a role for managing the PowerDNS Authoritative
Server. Configuration is based on the principles of Universal
Configuration and has been made explicitly forward-compatible up to
PowerDNS 4.5. Currently the role installs PowerDNS from the Debian
repositories, configures the gpgsql (PostgreSQL) backend by default and
performs a few overrides that make pdns more friendly to the DebOps
userbase:

- DNSSEC is automatically enabled for the gpgsql backend, but needs
  additional per-domain metadata to function. Hopefully this makes
  deploying DNSSEC a little more attractive.
- The IP address of the internal stub resolver is set automatically.
- The API key is created automatically when the API is enabled.
- The HTTP port is set to 16836 instead of 8081, which is reserved in
  Debian's /etc/services file.
- If the embedded webserver (for metrics and API usage) is enabled, a
  local NGINX server will be configured as a reverse proxy so that the
  webserver can be accessed securely over TLS.
- Dynamic DNS Update (RFC 2136) is automatically enabled, but needs
  additional per-domain configuration in the form of domain metadata to
  function. This makes pdns a little easier to use together with dhcpd.

imrejonk committed:
[pdns] Add new role debops.pdns

This adds debops.pdns, a role for managing the PowerDNS Authoritative
Server. Configuration is based on the principles of Universal
Configuration and has been made explicitly forward-compatible up to
PowerDNS 4.5. Currently the role installs PowerDNS from the Debian
repositories, configures the gpgsql (PostgreSQL) backend by default and
performs a few overrides that make pdns more friendly to the DebOps
userbase:

- DNSSEC is automatically enabled for the gpgsql backend, but needs
  additional per-domain metadata to function. Hopefully this makes
  deploying DNSSEC a little more attractive.
- The IP address of the internal stub resolver is set automatically.
- The API key is created automatically when the API is enabled.
- The HTTP port is set to 16836 instead of 8081, which is reserved in
  Debian's /etc/services file.
- If the embedded webserver (for metrics and API usage) is enabled, a
  local NGINX server will be configured as a reverse proxy so that the
  webserver can be accessed securely over TLS.
- Dynamic DNS Update (RFC 2136) is automatically enabled, but needs
  additional per-domain configuration in the form of domain metadata to
  function. This makes pdns a little easier to use together with dhcpd.

imrejonk commented:
This is currently a draft pull request; I'm comfortable with a merge after I've taken this on the production ride, which will probably be next week. Any preliminary feedback is of course greatly appreciated.

imrejonk commented:
I should note that I have only tested this role on Bullseye LXC containers so far. Our production authoritative nameservers run on Buster, so I might have to make some adjustments for that. We'll see.

imrejonk commented:
Another thing to note: I'm importing the playbook from `srv/all.yml` instead of `net/all.yml`. The reason for this is that pdns relies on a backend to function, and the default backend I'm configuring with this role is PostgreSQL Server, whose playbook is imported earlier in `srv/all.yml`. The same goes for the NGINX playbook (needed for the HTTP reverse proxy).

imrejonk committed:
[pdns] Improve role defaults

- Only configure settings if necessary.
- Add backwards compatibility with pdns < 4.2.0.

imrejonk committed:
[pdns] Add meta/main.yml file

imrejonk commented:
The production run went fine :rocket: 

@drybjed is it okay if I set the `min_ansible_version` tag in `meta/main.yml` to `0.0.0`? I don't actually know what the minimum Ansible version for this role would be, I've only tested it with 2.10 (and don't plan to test it with anything older).

imrejonk committed:
[pdns] Configure 'setgid' and 'setuid'

This ensures that PowerDNS is not running as root. Don't know how I
managed to miss this in the default `/etc/powerdns/pdns.conf`...

ypid commented on ansible/roles/pdns/tasks/init_postgresql.yml:
```diff
@@ -0,0 +1,21 @@
+    name: '{{ pdns__postgresql_database }}'
+    target: '{{ pdns__postgresql_schema }}'
+    state: 'restore'
+  when: pdns__register_postgresql_status is changed  # noqa no-handler
```
This file looks like a job for `debops.postgresql`. I am not sure if schema import is supported, but the role could be extended.

imrejonk replied:
`debops.postgresql` doesn't support schema imports, sadly. I'll take a look at extending it, but it's a bit outside the scope of this PR so I'll open a new one. @drybjed can decide whether the `debops.pdns` role can be merged as-is or if we should do the `debops.postgresql` fix first.

ypid commented on ansible/playbooks/service/pdns.yml:
```diff
@@ -0,0 +1,40 @@
+    - role: ferm
+      tags: [ 'role::ferm', 'skip::ferm' ]
+      ferm__dependent_rules:
+        - '{{ pdns__ferm__dependent_rules }}'
```
For example nginx dependent_rules are missing here. I guess with this design of role dependencies on the playbook level, you would have to check all roles what dependencies they have.

imrejonk replied:
Right, I never thought about this. I see that it's done like this with `debops.owncloud` as well. The downside of this approach is that it basically makes copies of the `service/nginx` playbook. Wouldn't it be cleaner to just instruct the user to add the host to the `[debops_service_nginx]` inventory group as well? Just to prevent code duplication and to keep some separation between the roles?

ypid replied:
This is a tradeoff between maintenance overhead and runtime. I guess we favor runtime but you are right. I am not fully happy with the duplication either.

imrejonk replied:
Addressed in e938bf0f5c36d5c4fa8f21f41f7beec0bfeb9a6f

ypid reviewed (changes requested):
I had a quick look out of interest. Overall it looks very good.

> I've only tested it with 2.10 (and don't plan to test it with anything older).

Debian 11 ships with Ansible 2.10 so that it not an unreasonable version to ask for. I would be fine when you set `min_ansible_version` to 2.10 to document that the role is fully supported starting with 2.10.

imrejonk committed:
[pdns] Set min_ansible_version to 2.10.0

Alphix commented on docs/ansible/roles/pdns/getting-started.rst:
```diff
@@ -0,0 +1,125 @@
+~~~~~~~~~~~~~~~~~~~
+
+Secondary operation can be enabled by setting :envvar:`pdns__secondary` to
+``True``. Doing so will instruct pdns to periodically check for zone changes at the primary nameservers, and update the local zones accordingly. These checks happen every 'refresh' seconds (as specified by the SOA record) and are only performed for domains with type SLAVE in your backend. Additionally, if the primary
```
Weird formatting?

imrejonk replied:
Thanks for pointing this out, fixed.

imrejonk committed:
[docs] Fix line wrapping in pdns documentation

imrejonk committed:
[pdns] Split playbook in 'plain' and 'nginx' parts

For hosts added to the `debops_service_pdns` group, only the PowerDNS
Authoritative Server and dependent roles are executed. When a host is
added to the `debops_service_pdns_nginx` group, the playbook will
additionally install Nginx and configure a reverse proxy for the pdns
API and metrics endpoint.

drybjed reviewed (approved):
Excellent work, thanks! I'll update the configuration template a bit to allow comments in YAML text blocks, but otherwise, looks good. :-)
